[AIRFLOW-2855] Check Cron Expression Validity in DagBag.process_file()

XD-DENG opened the pull request:
### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-2855
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

`schedule_interval` of DAGs can either be timedelta or a Cron expression.

When it's a Cron expression, there is no mechanism to check its validity at this moment. If there is anything wrong with the Cron expression itself, it will cause issues when methods `following_schedule()`  and `previous_schedule()` are invoked (will affect scheduling).

However, exceptions will only be written into logs. From Web UI, it’s hard for users to identify this issue & the source while no new task can be initiated (especially for users who’re not very familiar with Cron).

 It may be good to show error messages in web UI when a DAG's Cron expression (as schedule_interval) can not be parsed by `croniter` properly (this is implemented by adding these exceptions into metadata `import_error`, whose entries will be shown as error messages in web UI).

Related tests are added as well.

**Screenshot**
<img width="1680" alt="screen shot 2018-08-05 at 10 23 46 pm" src="https://user-images.githubusercontent.com/11539188/43687000-0ced5a5e-9901-11e8-85ba-b1d8ce0fce9f.png">



### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

KevinYang21 commented on airflow/models.py:
```diff
@@ -414,6 +427,16 @@ def process_file(self, filepath, only_if_updated=True, safe_mode=True):
+                        self.import_errors[dag.full_filepath] = \
+                            "Invalid Cron expression: " + str(cron_e)
+                        self.file_last_changed[dag.full_filepath] = \
+                            file_last_changed_on_disk
```
I may be missing something but do we need to perform the check twice?

XD-DENG replied:
One is for `.py` DAG file, and the another is for package DAG file (`.zip` file).

KevinYang21 replied:
So we append checked `m` to `mods` here, https://github.com/apache/incubator-airflow/pull/3698/files#diff-a32a363fa616685db3bfefba947535b2R370, and the if statement evaluating zip file or not ends here https://github.com/apache/incubator-airflow/pull/3698/files#diff-a32a363fa616685db3bfefba947535b2R416. So I think the `mods` we are iterating here, https://github.com/apache/incubator-airflow/pull/3698/files#diff-a32a363fa616685db3bfefba947535b2R418, is a combination of modules from `.py` file and modules from zip files.

XD-DENG replied:
Hi @yrqls21 , you're right. Thanks for pointing this out.

I'll modify accordingly.

KevinYang21 commented on tests/models.py:
```diff
@@ -1038,6 +1038,21 @@ def test_zip(self):
         dagbag.process_file(os.path.join(TEST_DAGS_FOLDER, "test_zip.zip"))
         self.assertTrue(dagbag.get_dag("test_zip_dag"))
 
+    def test_process_file_cron_validity_check(self):
```
I guess we do still want to keep the file in 644 instead of 755 mode.

KevinYang21 commented on tests/models.py:
```diff
@@ -1038,6 +1038,21 @@ def test_zip(self):
+            dagbag.process_file(os.path.join(TEST_DAGS_FOLDER, d))
+
+        for d in invalid_dag_files:
+            self.assertTrue(any([d in k and "Invalid Cron expression" in v
```
NIT: I think we don't need `any`

XD-DENG replied:
The statement inside `any()` is returning a List of Boolean values.
`Any()` helps check if there is at least one `True` element.

KevinYang21 replied:
Are we expecting more than one import error from importing the dedicated test DAG file?

XD-DENG replied:
`dagbag.import_errors` is a Dict whose Key are file path and Value are exception details.

Here what I did is to check "File is the one I'm intending to check" AND "exception is the exception that I'm looking into". The Length of this Boolean value List is the same as the number of entries in metadata `import_error`, in which there may be other errors other than those I'm checking here (here I only try to identify Cron invalidity errors.).

KevinYang21 replied:
Sry I missed the part that you are parsing the default DAG_HOME-indeed there might be other import errors. Would you double check what would be processed by calling models.DagBag() please? If `"test_invalid_cron.py", "test_zip_invalid_cron.zip"`are included I guess we don't need to reprocess them and if not I guess we don't need to parse the default DAG_HOME at all.

Back to the original point, what originally made me feel a bit strange is that we can actually check the errors in `O(n)` but now we do that in `O(n^m), m=# of import errors`. But it practically makes no difference so that's why it is a NIT, it's your freedom to give or take.

XD-DENG replied:
Hi @yrqls21 , great points!

- For your **1st point**, actually you're right. It's not really necessary to re-parse these two files, i.e. the code below can be moved.

```
for d in invalid_dag_files:
    dagbag.process_file(os.path.join(TEST_DAGS_FOLDER, d))
``` 
But I purposely keep them here to make it clear to readers what I'm testing here.

- Your **2nd point** is great. I agree it makes no practical difference here, given `n` and `m` would never be too big here. But I think I will change this part as well. Always good to keep performance in mind.

Thanks for your reviewing!

KevinYang21 commented on airflow/models.py:
```diff
@@ -414,6 +416,16 @@ def process_file(self, filepath, only_if_updated=True, safe_mode=True):
                         self.bag_dag(dag, parent_dag=dag, root_dag=dag)
                         found_dags.append(dag)
                         found_dags += dag.subdags
+                        if isinstance(dag._schedule_interval, six.string_types):
```
maybe put the check before we append the `dag` to `found_dags`?( if that's what we do for other dags with import errors)

KevinYang21 commented on tests/models.py:
```diff
@@ -1038,6 +1038,23 @@ def test_zip(self):
+        as schedule interval can be identified
+        """
+        invalid_dag_files = ["test_invalid_cron.py", "test_zip_invalid_cron.zip"]
+        dagbag = models.DagBag()
```
NIT: I understand that you want your test to show clearly that you are processing those two test files so it make sense to keep the `dagbag.process_file` line. Then I guess we can here process something like an empty/made up directory--the test DAG files in the default DAG_HOME may grow and then we have 2x parsing time growth for the additional file, and our CI is already painfully lengthy ;)

KevinYang21 commented on tests/models.py:
```diff
@@ -1038,6 +1038,23 @@ def test_zip(self):
+                                 for k, v in dagbag.import_errors.items()
+                                 if "Invalid Cron expression" in v]
+        for d in invalid_dag_files:
+            self.assertTrue(d in files_with_cron_error)
```
Sry for being picky here, just NIT, it is still `O(n^2)` here since `files_with_cron_error ` is a list :P

XD-DENG replied:
@yrqls21 come on, `n` is 2 here LOL

But I do agree with your another point on the `DagBag`'s `dag_folder`. What I have done in my latest commit is to pass a temporary directory to `dag_folder` of `DagBag`, so that it will not go scan all the test DAGs.

In addition, changing `dag_folder` of `DagBag` to a temp directory also makes the testing much easier. We simply need to check whether the number of errors grows from `0` to `# of invalid Cron tests`. No more big `O` LOL!

I do love your reviews! All valid and helpful points. Thanks!

KevinYang21 replied:
LOL I was just messing around for the holy perfectionism spirit. We can even be more assertive if we know those two are import error from cron exception. Good job for making the change to improve Airflow! Thank you!

(it's still `O(n)` tho ;))

XD-DENG replied:
:-|   Crazy now LOL

Thanks again! And look forward to your reviewing in any of my future commits!

KevinYang21 reviewed (approved):
LGTM besides the NITs, summoning committers @Fokko @ashb @kaxil @aoen

feng-tao commented on tests/models.py:
```diff
@@ -56,7 +56,7 @@
 from mock import patch, ANY
 from parameterized import parameterized
-from tempfile import NamedTemporaryFile
+from tempfile import NamedTemporaryFile, mkdtemp
```
small nit: could you move mkdtemp before NamedTemporaryFile given we put lower case import before upper case(L57)?

XD-DENG replied:
Updated as suggested. Thanks.

feng-tao reviewed (approved):
overall lgtm

XD-DENG committed:
[AIRFLOW-2855] Check Cron Expression Validity in DagBag.process_file()

A DAG can be imported as a .py script properly,
but the Cron expression inside as "schedule_interval" may be
invalid, like "0 100 * * *".

This commit helps check the validity of Cron expression in DAG
files (.py) and packaged DAG files (.zip), and help show
exception messages in web UI by add these exceptions into
metadata "import_error".
